File: packages/core/src/catalog/resolve.ts
/**
 * Parameter resolution: raw document values -> numbers, booleans and symbolic
 * forms, driven by the block's declared parameter specs.
 *
 * Type direction matters: without the spec we could not tell the enum `"silu"`
 * from the expression `"D"`.
 */

import type { ParamValue, Resolved, SymbolTable } from "../ir/types.js";
import type { BlockDef, ParamSpec, Ports, PortsSpec } from "./types.js";
import { Sym } from "../shapes/symexpr.js";
import { evalExpr } from "../shapes/expr.js";
import { symbolCtx } from "../ir/symbols.js";

function specOf(def: BlockDef, key: string): ParamSpec | undefined {
  return (def.params as Record<string, ParamSpec>)[key];
}

export function resolveNodeParams(
  def: BlockDef,
  raw: Record<string, ParamValue> | undefined,
  symbols: SymbolTable,
): Resolved {
  const ctx = symbolCtx(symbols);
  const out: Resolved = { type: def.type, p: {}, s: {}, raw: raw ?? {}, rawFull: {}, errors: [] };

  const keys = new Set<string>([...Object.keys(def.params ?? {}), ...Object.keys(raw ?? {})]);

  for (const key of keys) {
    const spec = specOf(def, key);
    const given = raw?.[key];

    if (!spec) {
      out.errors.push(`Unknown parameter "${key}" on block "${def.type}"`);
      out.p[key] = given;
      continue;
    }

    const value = given === undefined ? (spec as { default?: ParamValue }).default : given;
    if (value !== undefined) out.rawFull[key] = value;

    if (value === undefined) {
      if (spec.type === "bool") out.p[key] = false;
      else if (spec.type === "obj") out.p[key] = null;
      else out.errors.push(`Missing required parameter "${key}" on block "${def.type}"`);
      continue;
    }

    switch (spec.type) {
      case "int":
      case "num": {
        if (typeof value === "number") {
          out.p[key] = value;
          out.s[key] = Sym.con(value);
        } else if (typeof value === "string") {
          try {
            const sym = evalExpr(value, ctx);
            const n = sym.toNumber(ctx.values);
            if (n === null) {
              out.errors.push(`Parameter "${key}" does not evaluate to a number: ${sym.toString()}`);
            } else {
              out.p[key] = n;
              out.s[key] = sym;
            }
          } catch (e) {
            out.errors.push(`Parameter "${key}": ${(e as Error).message}`);
          }
        } else if (value === null) {
          out.p[key] = null;
        } else {
          out.errors.push(`Parameter "${key}" should be a number or expression`);
        }
        if (spec.type === "int" && typeof out.p[key] === "number" && !Number.isInteger(out.p[key])) {
          out.errors.push(
            `Parameter "${key}" must be an integer but evaluates to ${out.p[key]}` +
              (typeof value === "string" ? ` (from "${value}")` : ""),
          );
        }
        if (typeof out.p[key] === "number") {
          const s = spec as { min?: number; max?: number };
          if (s.min !== undefined && out.p[key] < s.min) {
            out.errors.push(`Parameter "${key}" must be >= ${s.min}, got ${out.p[key]}`);
          }
          if (s.max !== undefined && out.p[key] > s.max) {
            out.errors.push(`Parameter "${key}" must be <= ${s.max}, got ${out.p[key]}`);
          }
        }
        break;
      }
      case "bool": {
        // `null` is kept rather than coerced: some booleans are tri-state, such
        // as an output-projection bias that defaults to whatever the other
        // projections use. Consumers test with `=== true`.
        if (typeof value === "boolean") out.p[key] = value;
        else if (value === null) out.p[key] = null;
        else out.errors.push(`Parameter "${key}" should be a boolean`);
        break;
      }
      case "enum": {
        if (typeof value === "string" && spec.values.includes(value)) out.p[key] = value;
        else if (value === null) out.p[key] = null;
        else out.errors.push(`Parameter "${key}" should be one of ${spec.values.join(", ")}, got ${String(value)}`);
        break;
      }
      case "str":
      case "pattern": {
        if (typeof value === "string") out.p[key] = value;
        else if (value === null) out.p[key] = null;
        else out.errors.push(`Parameter "${key}" should be a string`);
        break;
      }
      case "obj": {
        out.p[key] = value;
        break;
      }
    }
  }

  return out;
}

export function portsOf(spec: PortsSpec, r: Resolved): Ports {
  return typeof spec === "function" ? spec(r) : spec;
}

File: packages/core-go/catalog/resolve.go
package catalog

import (
	"fmt"
	"math"
	"sort"
	"strings"

	"github.com/tensorcad/core/attnexpr"
	"github.com/tensorcad/core/ir"
	"github.com/tensorcad/core/shapes"
)

// ResolveNodeParams turns the raw values a document wrote into numbers,
// booleans and symbolic forms, driven by the block's declared parameter specs.
//
// Type direction matters: without the spec there would be no way to tell the
// enum "silu" from the expression "D".
func ResolveNodeParams(def *BlockDef, raw map[string]any, symbols *ir.SymbolTable) *Resolved {
	ctx := ir.SymbolCtx(symbols)
	out := &Resolved{
		Type:    def.Type,
		P:       map[string]any{},
		S:       map[string]shapes.Sym{},
		Raw:     raw,
		RawFull: map[string]any{},
	}
	if out.Raw == nil {
		out.Raw = map[string]any{}
	}

	// Declared parameters first, in the order the block declares them, then
	// anything else the document wrote. The declared order is what a reader
	// sees; the extras are all errors, so they only need to be in *some* order,
	// and sorted is the one that does not change between runs.
	extra := map[string]bool{}
	for k := range raw {
		if !def.Params.Has(k) {
			extra[k] = true
		}
	}
	keys := make([]string, 0, len(def.Params)+len(extra))
	for _, e := range def.Params {
		keys = append(keys, e.Name)
	}
	keys = append(keys, sortedKeys(extra)...)
	out.Keys = keys

	for _, key := range keys {
		spec, hasSpec := def.Params.Get(key)
		given, hasGiven := raw[key]

		if !hasSpec {
			out.Errors = append(out.Errors,
				fmt.Sprintf("Unknown parameter %q on block %q", key, def.Type))
			out.P[key] = given
			continue
		}

		var value any
		var hasValue bool
		if hasGiven {
			value, hasValue = given, true
		} else if spec.HasDefault {
			value, hasValue = spec.Default, true
		}
		if hasValue {
			out.RawFull[key] = value
		}

		if !hasValue {
			switch spec.Type {
			case ParamBool:
				out.P[key] = false
			case ParamObj:
				out.P[key] = nil
			default:
				out.Errors = append(out.Errors,
					fmt.Sprintf("Missing required parameter %q on block %q", key, def.Type))
			}
			continue
		}

		switch spec.Type {
		case ParamInt, ParamNum:
			resolveNumeric(out, key, spec, value, ctx)
		case ParamBool:
			// A null is kept rather than coerced: some booleans are tri-state,
			// such as an output-projection bias that defaults to whatever the
			// other projections use. Consumers test for true explicitly.
			switch v := value.(type) {
			case bool:
				out.P[key] = v
			case nil:
				out.P[key] = nil
			default:
				out.Errors = append(out.Errors, fmt.Sprintf("Parameter %q should be a boolean", key))
			}
		case ParamEnum:
			switch v := value.(type) {
			case string:
				if contains(spec.Values, v) {
					out.P[key] = v
				} else {
					out.Errors = append(out.Errors, fmt.Sprintf(
						"Parameter %q should be one of %s, got %v", key, join(spec.Values), v))
				}
			case nil:
				out.P[key] = nil
			default:
				out.Errors = append(out.Errors, fmt.Sprintf(
					"Parameter %q should be one of %s, got %v", key, join(spec.Values), value))
			}
		case ParamStr, ParamPattern:
			switch v := value.(type) {
			case string:
				out.P[key] = v
			case nil:
				out.P[key] = nil
			default:
				out.Errors = append(out.Errors, fmt.Sprintf("Parameter %q should be a string", key))
			}
		case ParamObj:
			out.P[key] = value
		case ParamMask, ParamScore:
			resolveExpression(out, key, spec.Type, value, symbols)
		}
	}
	return out
}

// resolveExpression compiles an attention expression against the design's
// symbols and keeps it in the form it was understood in: every symbol replaced
// by its value, every constant folded, and only the parentheses it needs. That
// is what the analysis evaluates and the generated code prints, so it is what
// a reader is shown as the evaluated value.
//
// Nothing is stored for an empty expression, or for one that did not compile:
// a composite passes on what it resolved, and an error reported here would
// otherwise be reported again by every block it expands into.
//
// B and T are left out. They carry a default in the symbol table, but a mask
// that read T would be evaluated at that default rather than at the length it
// runs at, and the analysis measures it at the operating point's.
func resolveExpression(out *Resolved, key string, kind ParamKind, value any, symbols *ir.SymbolTable) {
	text, ok := value.(string)
	if value != nil && !ok {
		out.Errors = append(out.Errors, fmt.Sprintf("Parameter %q should be an expression", key))
		return
	}
	if strings.TrimSpace(text) == "" {
		return
	}
	want := attnexpr.Mask
	if kind == ParamScore {
		want = attnexpr.Score
	}
	values := map[string]float64{}
	if symbols != nil {
		for name, v := range symbols.Values {
			if !symbols.Runtime[name] {
				values[name] = v
			}
		}
	}
	n, err := attnexpr.Compile(text, want, values)
	if err != nil {
		out.Errors = append(out.Errors, fmt.Sprintf("Parameter %q: %s", key, err))
		return
	}
	out.P[key] = attnexpr.String(n)
}

func resolveNumeric(out *Resolved, key string, spec ParamSpec, value any, ctx shapes.EvalCtx) {
	switch v := value.(type) {
	case float64:
		out.P[key] = v
		out.S[key] = shapes.Con(v)
	case int:
		out.P[key] = float64(v)
		out.S[key] = shapes.Con(float64(v))
	case string:
		sym, err := shapes.EvalExpr(v, ctx)
		if err != nil {
			out.Errors = append(out.Errors, fmt.Sprintf("Parameter %q: %s", key, err))
			return
		}
		// Evaluate rather than ToNumber: a symbol that failed carries NaN, and a
		// block written over it has to report that rather than fault. The
		// editor is where this lands, and a faulted engine cannot be restarted
		// without reloading the window.
		n, ok, err := sym.Evaluate(ctx.Values)
		if err != nil {
			out.Errors = append(out.Errors, fmt.Sprintf("Parameter %q: %s", key, err))
			return
		}
		if !ok {
			out.Errors = append(out.Errors, fmt.Sprintf(
				"Parameter %q does not evaluate to a number: %s", key, sym))
			return
		}
		out.P[key] = n
		out.S[key] = sym
	case nil:
		out.P[key] = nil
		return
	default:
		out.Errors = append(out.Errors,
			fmt.Sprintf("Parameter %q should be a number or expression", key))
		return
	}

	n, isNum := out.P[key].(float64)
	if !isNum {
		return
	}
	if spec.Type == ParamInt && n != math.Trunc(n) {
		msg := fmt.Sprintf("Parameter %q must be an integer but evaluates to %s", key, num(n))
		if s, ok := value.(string); ok {
			msg += fmt.Sprintf(" (from %q)", s)
		}
		out.Errors = append(out.Errors, msg)
	}
	if spec.Min != nil && n < *spec.Min {
		out.Errors = append(out.Errors,
			fmt.Sprintf("Parameter %q must be >= %s, got %s", key, num(*spec.Min), num(n)))
	}
	if spec.Max != nil && n > *spec.Max {
		out.Errors = append(out.Errors,
			fmt.Sprintf("Parameter %q must be <= %s, got %s", key, num(*spec.Max), num(n)))
	}
}

// num formats a float the way JavaScript prints one, because these strings end
// up in messages the two engines have to agree on.
func num(v float64) string {
	if v == math.Trunc(v) && math.Abs(v) < 1e21 {
		return fmt.Sprintf("%d", int64(v))
	}
	return fmt.Sprintf("%v", v)
}

func contains(list []string, v string) bool {
	for _, s := range list {
		if s == v {
			return true
		}
	}
	return false
}

func join(list []string) string {
	out := ""
	for i, s := range list {
		if i > 0 {
			out += ", "
		}
		out += s
	}
	return out
}

func sortedKeys(set map[string]bool) []string {
	out := make([]string, 0, len(set))
	for k := range set {
		out = append(out, k)
	}
	sort.Strings(out)
	return out
}
